pennylane/beta/tapes/tape.py

```diff
@@ -833,3 +835,105 @@ def data(self):
+
+        **Example**
+
+        .. code-block:: python
+
+            from pennylane.beta.tapes import QuantumTape
+            from pennylane.beta.queuing import expval, var, sample, probs
+
+            with QuantumTape() as tape:
+                qml.RX(0.432, wires=0)
+                qml.RY(0.543, wires=0)
+                qml.CNOT(wires=[0, 'a'])
+                qml.RX(0.133, wires='a')
+                probs(wires=[0, 'a'])
+
+        If parameters are not provided, the existing tape parameters are used:
+
+        >>> dev = qml.device("default.qubit", wires=[0, 'a'])
+        >>> tape.execute(dev)
+        array([[8.84828969e-01, 3.92449987e-03, 4.91235209e-04, 1.10755296e-01]])
+
+        Parameters can be optionally passed during execution:
+
+        >>> tape.execute(dev, params=[1.0, 0.0, 1.0])
+        array([[0.5931328 , 0.17701835, 0.05283049, 0.17701835]])
+
+        Parameters provided for execution are temporary, and do not affect
+        the tapes' parameters in-place:
+
+        >>> tape.get_parameters()
+        [0.432, 0.543, 0.133]
+        """
+        if params is None:
+            params = self.get_parameters()
+
+        return self._execute(params, device=device)
+
+    def execute_device(self, params, device):
+        """Execute the tape on a quantum device.
+
+        This is a low-level method, intended to be called by an interface,
+        and does not support autodifferentiation.
+
+        For more details on differentiable tape execution, see :meth:`~.execute`.
+
+        Args:
+            device (~.Device): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+        """
+        device.reset()
+
+        # backup the current parameters
+        current_parameters = self.get_parameters()
+
+        # temporarily mutate the in-place parameters
+        self.set_parameters(params)
+
+        if isinstance(device, qml.QubitDevice):
+            res = device.execute(self)
+        else:
+            res = device.execute(self.operations, self.observables, {})
+
+        # Update output dim if incorrect.
+        # Note that we cannot assume the type of `res`, so
+        # we use duck typing to catch any 'array like' object.
+        try:
+            if isinstance(res, np.ndarray) and res.dtype is np.dtype("object"):
+                output_dim = sum([len(i) for i in res])
+            else:
+                output_dim = np.prod(res.shape)
+
+            if self.output_dim != output_dim:
+                # update the output dimension estimate with the correct value
+                self._output_dim = output_dim
+
+        except (AttributeError, TypeError):
+            # unable to determine the output dimension
+            pass
```

Comment:
codecov seems to have a legitimate complaint here

Reply:
Woah! This in-line commenting by codecov is new